Update specifications ADDRESS02 (#1293)
dotse/zonemaster: docs/public/specifications/tests/Address-TP/address02.md

severity level can be changed in the [Zonemaster-Engine profile]. Also see the
[Severity Level Definitions] document.

The argument names in the Arguments column lists the arguments used in the
message. The argument names are defined in the [argument list].

## Test procedure 

1. Create the empty set: Name server name and IP address ("Name Server IP").

2. Obtain the glue address records of each name server for the domain from the
   parent using the method [Get-Del-NS-Names-and-IPs] and add them to the 
   *Name Server IP* set. 

3. Obtain the IP addresses of each name server for the domain using the method 
   [Get-Zone-NS-Names-and-IPs] and add any non-duplicate results to 
   *Name Server IP* set. 

4. Create the following empty set: IP address ("PTR Missing")

5. For each name server in *Name Server IP* do:
   1. Make a recursive PTR query.
   2. If the response fails to match the following criteria, add the IP address
      to the *PTR Missing* set.
        - RCODE must be NOERROR
        - answer section must contain at least one PTR record
  
6. If the set *PTR Missing* is empty, then putput *[A02_PTR_PRESENT]*

> 7. Else, output *[A02_PTR_MISSING]* with a list of the IP addresses in the set.

Reviewer: I suggest

`(...) list of the IP addresses in the  *PTR Missing* set.`

> # ADDRESS02: Reverse DNS entry exists for name server IP address

Reviewer: The head line is quite long. What do you think of the following, shorter alternatives?
* Reverse DNS entry for name server IP
* PTR exists for name server IP

## Test case identifier
**ADDRESS02** 

## Table of contents

* [Objective](#objective)
* [Scope](#scope)
* [Inputs](#inputs)
* [Summary](#summary)
* [Test procedure](#test-procedure)
* [Outcome(s)](#outcomes)
* [Special procedural requirements](#special-procedural-requirements)
* [Intercase dependencies](#intercase-dependencies)

## Objective

Best curent practices dictates that imternet reachable hosts should have a
reverse DNS entry, as various services on the Internet, for instance spam 
filters, may consider this when determining the trustworthiness of the host.
Se [RFC1912] section 2.1 and [RFC1033] page 11 for additional information.

## Scope

This test checks for the existence of PTR records for the corresponding reverse

> domains of the name servers IP address.

Reviewer: This test checks for the existence of PTR records for the corresponding reverse
domains of the name servers IP address.

## Inputs*

The domain name to be tested.

## Summary

Message Tag                   | Level    | Arguments | Message ID for message tag
:---------------------------- |:---------|:----------|:--------------------------

> A02_PTR_PRESENT               | INFO     |           | PTR present for all nameservers

Reviewer: Suggestion:
"A PTR record is present for every name server IP address."

> A02_PTR_MISSING               | NOTICE   | ns_list   | PTR missing for "{ns_list}"

Reviewer: Suggestion:
```
PTR record is missing for the following name server IP addresses: "{ns_list}"
```

The value in the Level column is the default severity level of the message. The
severity level can be changed in the [Zonemaster-Engine profile]. Also see the
[Severity Level Definitions] document.

The argument names in the Arguments column lists the arguments used in the
message. The argument names are defined in the [argument list].

## Test procedure 

1. Create the empty set: Name server name and IP address ("Name Server IP").

2. Obtain the glue address records of each name server for the domain from the
   parent using the method [Get-Del-NS-Names-and-IPs] and add them to the 

>    *Name Server IP* set. 

Reviewer: With the method we get all name server names in the delegation and all IP addresses for those names, but the IP addresses do not have to be from glue records. They could be out-of-bailiwick and fetched seperately.

Compare with the update in https://github.com/zonemaster/zonemaster/pull/1284

3. Obtain the IP addresses of each name server for the domain using the method 
   [Get-Zone-NS-Names-and-IPs] and add any non-duplicate results to 
   *Name Server IP* set. 

4. Create the following empty set: IP address ("PTR Missing")

5. For each name server in *Name Server IP* do:
   1. Make a recursive PTR query.
   2. If the response fails to match the following criteria, add the IP address
      to the *PTR Missing* set.
        - RCODE must be NOERROR
        - answer section must contain at least one PTR record
  
6. If the set *PTR Missing* is empty, then putput *[A02_PTR_PRESENT]*

7. Else, output *[A02_PTR_MISSING]* with a list of the IP addresses in the 
   *PTR Missing* set.

## Outcome(s)

The outcome of the Test Case is "notice" if there is at least one message with
the message with the severity level *[NOTICE]*.

If there are no message or only a message with the severity level *[INFO]*, the

> outcome of the Test Case is "pass".

Reviewer: Make the Outcome section identical to the Outcome section in #1284.

Also, make sure that `[NOTICE]` etc are in the link list.

> # ADDRESS02: PTR exists for name server IP

Reviewer 2: Unsure if this is better to change. Maybe this?
```suggestion
# ADDRESS02: Reverse entry (PTR) exists for name server IP address
```

## Test case identifier
**ADDRESS02** 

## Table of contents

* [Objective](#objective)
* [Scope](#scope)
* [Inputs](#inputs)
* [Summary](#summary)
* [Test procedure](#test-procedure)
* [Outcome(s)](#outcomes)
* [Special procedural requirements](#special-procedural-requirements)
* [Intercase dependencies](#intercase-dependencies)

## Objective

> Best curent practices dictates that internet reachable hosts should have a

Reviewer 2: ```suggestion
Best curent practices dictate that internet reachable hosts should have a
```

> reverse DNS entry, as various services on the Internet, for instance spam 
> filters, may consider this when determining the trustworthiness of the host.

Reviewer 2: ```suggestion
reverse DNS entry, as various services on the Internet (e.g. spam 
filters) may consider this when determining the trustworthiness of the host.
```

See [RFC1912] section 2.1 and [RFC1033] page 11 for additional information.

This test checks for the existence of PTR records for the corresponding reverse
domains of the name servers IP address.

## Scope
Only the existence of a PTR record, or a record that resolves to a PTR record, 
is checked. Not the validity of said record. That is handeled by other tests.

## Inputs

> The domain name to be tested.

Reviewer: I suggest

```
* "Child Zone" -- the domain name to be tested.
```

Below there are references to `*Child Zone*` and then it should be defined first. See the wording in #1284.

> is checked. Not the validity of said record. That is handled by other tests.

Reviewer 2: ```suggestion
is checked. Not the validity of said record, which is handled by other tests.
```

> is checked. Not the validity of said record. That is handeled by other tests.

Reviewer 3: ```suggestion
is checked. Not the validity of said record. That is handled by other tests.
```

## Summary

> Message Tag                   | Level    | Arguments | Message ID for message tag
> :---------------------------- |:---------|:----------|:--------------------------
> A02_PTR_PRESENT               | INFO     |            | PTR record present for each nameserver IP address
> A02_PTR_MISSING               | NOTICE   | ns_ip_list | PTR missing for the following name server IP addresses: "{ns_ip_list}"

Reviewer 2: ```suggestion
Message Tag                   | Level    | Arguments  | Message ID for message tag
:---------------------------- |:---------|:-----------|:--------------------------
A02_PTR_PRESENT               | INFO     |            | PTR record present for all name server IP addresses
A02_PTR_MISSING               | NOTICE   | ns_ip_list | PTR record missing for the following name server IP addresses: "{ns_ip_list}"
```

The value in the Level column is the default severity level of the message. The
severity level can be changed in the [Zonemaster-Engine profile]. Also see the
[Severity Level Definitions] document.

The argument names in the Arguments column lists the arguments used in the
message. The argument names are defined in the [argument list].

> 

Reviewer 2: ```suggestion
The name server names are assumed to be available at the time when the msgid is created, if the argument name is "ns" or "ns_list" even when in the [Test procedure] below it is only referred to the IP address of the name servers.

```

## Test procedure 

> 1. Create the empty set: Name server name and IP address ("Name Server IP").
> 
> 2. Obtain the address records of each name server for the *Child Zone* from the
>    parent using the method [Get-Del-NS-Names-and-IPs] and add them to the 
>    *Name Server IP* set. 
> 
> 3. Obtain the IP addresses of each name server for the domain using the method 
>    [Get-Zone-NS-Names-and-IPs] and add any non-duplicate results to 
>    *Name Server IP* set. 

Reviewer 2: ```suggestion
1. Retrieve all name server names and IP addresses for *Child Zone* using
   methods [Get-Del-NS-Names-and-IPs] and [Get-Zone-NS-Names-and-IPs] (Name Server IP).
```

> 4. Create the following empty set: IP address ("PTR Missing")

Reviewer 2: ```suggestion
2. Create the following empty set: name server IP address ("PTR Missing")
```

5. For each name server in *Name Server IP* do:
   1. Make a recursive PTR query.
   2. If the response fails to match the following criteria, add the IP address
      to the *PTR Missing* set.
        - RCODE must be NOERROR

>         - answer section must contain at least one PTR record

Reviewer 2: ```suggestion
3. For each name server IP address in *Name Server IP* do:
   1. Do a reverse *DNS Lookup* (PTR) of name server IP address.
   2. If the response fails to match the following criteria, then
      add name server IP address to the *PTR Missing* set:
        - [RCODE Name] is "NOERROR"
        - Answer section contains at least one PTR record
```

> 6. If the set *PTR Missing* is empty, then putput *[A02_PTR_PRESENT]*

Reviewer 3: ```suggestion
6. If the set *PTR Missing* is empty, then output *[A02_PTR_PRESENT]*.
```

> 6. If the set *PTR Missing* is empty, then output *[A02_PTR_PRESENT]*

Reviewer 2: ```suggestion
6. If the set *PTR Missing* is empty, then output *[A02_PTR_PRESENT]*.
```

7. Else, output *[A02_PTR_MISSING]* with a list of the IP addresses in the 
   *PTR Missing* set.

## Outcome(s)

The outcome of this Test Case is "fail" if there is at least one message
with the severity level *[ERROR]* or *[CRITICAL]*.

The outcome of this Test Case is "warning" if there is at least one message
with the severity level *[WARNING]*, but no message with severity level 
*[ERROR]* or *[CRITICAL]*.

In other cases, no message or only messages with severity level
*[INFO]* or *[NOTICE]* the outcome of this Test Case is "pass".

## Special procedural requirements

None.

## Intercase dependencies

None.

> 

Reviewer 2: ```suggestion

## Terminology

* "DNS Lookup" - The term is used when a recursive lookup is used, though
any changes to the DNS tree introduced by an [undelegated test] must be
respected.

```

[A02_PTR_PRESENT]:                  #Summary
[A02_PTR_MISSING]:                  #Summary
[Argument list]:                    ../ArgumentsForTestCaseMessages.md
[CRITICAL]:                         ../SeverityLevelDefinitions.md#critical
[ERROR]:                             ../SeverityLevelDefinitions.md#error
[Get-Del-NS-Names-and-IPs]:         ../MethodsV2.md#method-get-delegation-ns-names-and-ip-addresses
[Get-Zone-NS-Names-and-IPs]:        ../MethodsV2.md#method-get-zone-ns-names-and-ip-addresses
[INFO]:                             ../SeverityLevelDefinitions.md#info
[NOTICE]:                             ../SeverityLevelDefinitions.md#notice

> [RFC1912]:                          https://www.rfc-editor.org/rfc/rfc1912

Reviewer 2: ```suggestion
[RCODE Name]:                       https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6
[RFC1912]:                          https://www.rfc-editor.org/rfc/rfc1912
```

[RFC1033]:                          https://www.rfc-editor.org/rfc/rfc1033
[Severity Level Definitions]:       ../SeverityLevelDefinitions.md

> [WARNING]:                          ../SeverityLevelDefinitions.md#warning

Reviewer 2: ```suggestion
[Test procedure]:                   #test-procedure
[Undelegated test]:                 ../../test-types/undelegated-test.md
[WARNING]:                          ../SeverityLevelDefinitions.md#warning
```
